lib/charms/tempo_k8s/v0/tracing.py

```diff
@@ -0,0 +1,485 @@
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        # ...
+        self.tracing = TracingEndpointProvider(self)
+        # ...
+
+Note that the first argument (`self`) to `TracingEndpointProvider` is always a reference to the
+parent charm.
+
+Units of provider charms obtain the tempo endpoint to which they will push their traces by using one 
+of these  `TracingEndpointProvider` attributes, depending on which protocol they support:
+- otlp_grpc_endpoint
+- otlp_http_endpoint
+- zipkin_endpoint
+- tempo_endpoint
+
+## Requirer Library Usage
+
+The `TracingEndpointRequirer` object may be used by charms to manage relations with their
+trace sources. For this purposes a Tempo-like charm needs to do two things
+
+1. Instantiate the `TracingEndpointRequirer` object by providing it a
+reference to the parent (Tempo) charm and optionally the name of the relation that the Tempo charm
+uses to interact with its trace sources. This relation must conform to the `tracing` interface
+and it is strongly recommended that this relation be named `tracing` which is its
+default value.
+
+For example a Tempo charm may instantiate the `TracingEndpointRequirer` in its constructor as
+follows
+
+    from charms.tempo_k8s.v0.tracing import TracingEndpointRequirer
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        # ...
+        self.tracing = TracingEndpointRequirer(self)
+        # ...
+
+
+
+"""  # noqa: W505
+import json
+import logging
+from typing import TYPE_CHECKING, List, Literal, MutableMapping, Optional, Tuple, cast
+
+import pydantic
+from ops.charm import CharmBase, CharmEvents, RelationEvent, RelationRole
+from ops.framework import EventSource, Object
+from ops.model import ModelError, Relation
+from pydantic import BaseModel
+
+# The unique Charmhub library identifier, never change it
+LIBID = "12977e9aa0b34367903d8afeb8c3d85d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 3
+
+PYDEPS = ["pydantic<2.0"]
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "tracing"
+RELATION_INTERFACE_NAME = "tracing"
+
+IngesterProtocol = Literal["otlp_grpc", "otlp_http", "zipkin", "tempo"]
+
+RawIngester = Tuple[IngesterProtocol, int]
+
+
+class TracingError(RuntimeError):
+    """Base class for custom errors raised by this library."""
+
+
+class DataValidationError(TracingError):
+    """Raised when data validation fails on IPU relation data."""
+
+
+class DatabagModel(BaseModel):
+    """Base databag model."""
+
+    _NEST_UNDER = None
+
+    @classmethod
+    def load(cls, databag: MutableMapping):
+        """Load this model from a Juju databag."""
+        if cls._NEST_UNDER:
+            return cls.parse_obj(json.loads(databag[cls._NEST_UNDER]))
+
+        data = {k: json.loads(v) for k, v in databag.items()}
```

Comment:
The unit test failure in the CI workflow is caused by this line. This method expects the values in the `databag` should all be valid JSON strings. Looks like we need to do `json.dumps` for the `host` value in the `databag`.

Reply:
do I need to json.dumps a string?I ll try and fix it

Reply:
```
>>> a = "asd"
>>> b = json.dumps(a)
>>> a
'asd'
>>> b
'"asd"'

```

ah mate @wood-push-melon this is new to me :+1:  thanks a lot